cleanlab/multilabel_classification/rank.py

```diff
@@ -0,0 +1,193 @@
+# Copyright (C) 2017-2023  Cleanlab Inc.
+# This file is part of cleanlab.
+#
+# cleanlab is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# cleanlab is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
+
+"""
+Methods to rank the severity of label issues in multi-label classification datasets.
+Here each example can belong to one or more classes, or none of the classes at all.
+Unlike in standard multi-class classification, predicted class probabilities from model need not sum to 1 for each row in multi-label classification.
+"""
+
+import numpy as np  # noqa: F401: Imported for type annotations
+import numpy.typing as npt
+from typing import List, TypeVar, Dict, Any, Optional, Tuple
+
+from cleanlab.internal.validation import assert_valid_inputs
+from cleanlab.internal.util import get_num_classes
+from cleanlab.internal.multilabel_utils import int2onehot
+from cleanlab.internal.multilabel_scorer import MultilabelScorer, ClassLabelScorer, Aggregator
+
+
+T = TypeVar("T", bound=npt.NBitBase)
+
+
+def _labels_to_binary(
+    labels: List[List[int]],
+    pred_probs: npt.NDArray["np.floating[T]"],
+) -> np.ndarray:
+    """Validate the inputs to the multilabel scorer. Also transform the labels to a binary representation."""
+    assert_valid_inputs(
+        X=None, y=labels, pred_probs=pred_probs, multi_label=True, allow_one_class=True
+    )
+    num_classes = get_num_classes(labels=labels, pred_probs=pred_probs, multi_label=True)
+    binary_labels = int2onehot(labels, K=num_classes)
+    return binary_labels
+
+
+def _create_multilabel_scorer(
+    method: str,
+    adjust_pred_probs: bool,
+    aggregator_kwargs: Optional[Dict[str, Any]] = None,
+) -> Tuple[MultilabelScorer, Dict]:
+    """This function acts as a factory that creates a MultilabelScorer."""
+    base_scorer = ClassLabelScorer.from_str(method)
+    base_scorer_kwargs = {"adjust_pred_probs": adjust_pred_probs}
+    if aggregator_kwargs:
+        aggregator = Aggregator(**aggregator_kwargs)
+        scorer = MultilabelScorer(base_scorer, aggregator)
+    else:
+        scorer = MultilabelScorer(base_scorer)
+    return scorer, base_scorer_kwargs
+
+
```

Comment:
Please make sure you also update the tutorial:
https://docs.cleanlab.ai/stable/tutorials/multilabel_classification.html

to properly import this function from here

Reply:
updated